lapce-data/src/data.rs

```diff
@@ -275,6 +309,68 @@ impl LapceData {
+        }
+        if let Some(str) = path.to_str() {
+            let mut splits = str.rsplit(':');
+            if let Some(first_rhs) = splits.next() {
+                if let Ok(first_rhs_num) = first_rhs.parse::<usize>() {
+                    // return if remainder after stripping first_rhs is filename and first_rhs is number
+                    if PathBuf::from(
+                        path.to_str().unwrap_or_default()
+                            .strip_suffix(first_rhs).unwrap_or_default()
+                            .strip_suffix(':').unwrap_or_default()
+                            .to_string()
+                    ).is_file() 
+                    {
+                        return Some((
+                            PathBuf::from(
+                                path.to_str().unwrap_or_default().strip_suffix(first_rhs).unwrap_or_default().strip_suffix(':').unwrap_or_default().to_string()
+                            ),
+                            LineCol { 
+                                line: first_rhs_num, 
+                                column: 1 
+                            }
+                        ))
+                    }
+                    if let Some(second_rhs) = splits.next() {
+                        if let Ok(second_rhs_num) = second_rhs.parse::<usize>() {
+                            // Return if remainder after stripping second_rhs is filename and second_rhs is number
+                            if PathBuf::from(
+                                path.to_str().unwrap_or_default()
+                                    .strip_suffix(first_rhs).unwrap_or_default()
+                                    .strip_suffix(':').unwrap_or_default()
+                                    .strip_suffix(second_rhs).unwrap_or_default()
+                                    .strip_suffix(':').unwrap_or_default()
+                                    .to_string()
+                            ).is_file() {
+                                return Some((
+                                    PathBuf::from(
+                                        path.to_str().unwrap_or_default()
+                                            .strip_suffix(first_rhs).unwrap_or_default()
+                                            .strip_suffix(':').unwrap_or_default()
+                                            .strip_suffix(second_rhs).unwrap_or_default()
+                                            .strip_suffix(':').unwrap_or_default()
+                                            .to_string()
+                                    ),
+                                    LineCol { 
+                                        line: second_rhs_num, 
+                                        column: first_rhs_num, 
+                                    }
+                                ))
+                            }      
+                        }
+                    }
+                }
+            }
+        }
+        None
+    }
```

Comment:
Let's just rewrite the function, without adding unnecessary complexity, introducing unnecessary restrictions, etc. Tests can be seen here (see below inside the main function): https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b20e680104335e7224aa8fbbd4f55f53
```rust
pub fn parse_path_line_column(path: &Path) -> Option<(PathBuf, LineCol)> {
    fn write_text_with_sep_to<'a, I>(mut iter: I, buf: &mut String, sep: &str)
    where
        I: Iterator<Item = &'a str>,
    {
        if let Some(str) = iter.next() {
            buf.push_str(str);
            buf.push_str(sep);
            write_text_with_sep_to(iter, buf, sep);
        }
    }
    if let Some(str) = path.to_str() {
        let mut splits = str.rsplit(':');
        if let Some(first_rhs) = splits.next() {
            match first_rhs.parse::<usize>() {
                Ok(first_rhs_num) => {
                    if let Some(second_rhs) = splits.next() {
                        match second_rhs.parse::<usize>() {
                            Ok(second_rhs_num) => {
                                let mut str = String::new();
                                write_text_with_sep_to(splits.rev(), &mut str, ":");

                                let left_path = PathBuf::from(&str[..&str.len() - 1]);

                                if left_path.is_file() {
                                    return Some((
                                        left_path,
                                        LineCol {
                                            line: second_rhs_num,
                                            column: first_rhs_num,
                                        },
                                    ));
                                }
                                // We have some second right hand number, but the remaining path isn't a file,
                                // then let's check if it changed if we add `second_rhs`?
                                // Last char of `str` is ":", so we neen to push only `second_rhs`
                                str.push_str(second_rhs);
                                let left_path = PathBuf::from(&str);
                                if left_path.is_file() {
                                    return Some((
                                        left_path,
                                        LineCol {
                                            line: first_rhs_num,
                                            column: 1,
                                        },
                                    ));
                                } else if path.is_file() {
                                    // Ok, that's not a file either. Then just check whole path
                                    return Some((
                                        path.to_path_buf(),
                                        LineCol { line: 1, column: 1 },
                                    ));
                                }
                            }
                            Err(_) => {
                                let mut str = String::new();
                                write_text_with_sep_to(splits.rev(), &mut str, ":");
                                str.push_str(second_rhs);

                                let left_path = PathBuf::from(&str);

                                if left_path.is_file() {
                                    return Some((
                                        left_path,
                                        LineCol {
                                            line: first_rhs_num,
                                            column: 1,
                                        },
                                    ));
                                } else if path.is_file() {
                                    // We have some number on the right but the remaining path isn't a file, then
                                    // let's check if the entire path is a file?
                                    return Some((
                                        path.to_path_buf(),
                                        LineCol { line: 1, column: 1 },
                                    ));
                                }
                            }
                        }
                    } else if path.is_file() {
                        return Some((path.to_path_buf(), LineCol { line: 1, column: 1 }));
                    }
                }
                Err(_) => {
                    // If the first right component isn't a number, then the enire path has to refer to a file name
                    if path.is_file() {
                        return Some((path.to_path_buf(), LineCol { line: 1, column: 1 }));
                    }
                }
            }
        }
    // Check if the path refers to a file if the path isn't UTF-8 valid
    } else if path.is_file() {
        return Some((path.to_path_buf(), LineCol { line: 1, column: 1 }));
    }
    // Ignore the path if it doesn't refer to a file
    None
}
```

Reply:
And advice about adding comments. I'm sure @panekj was pointing out over-commenting the obvious, but didn't ask for all comments to be removed. For example, this is obvious:
```rust
             // Path components separated by :
             let mut splits = str.rsplit(':');
           
             // Is the first right component a number?
             match first_rhs.parse::<usize>() {
```
But this is not obvious, so it's probably worth leaving a comment:
```rust
// Last char of `str` is ":", so we neen to push only `second_rhs`
str.push_str(second_rhs);
```
As a rule, you should not comment out functions from `std`, the **type** returned by the function, etc. This is already widely known or is already visible. It is worth commenting on the algorithm of work (if it is non-trivial), that is, not what the function return, but why it is needed here.